math/ncr_modulo_p.cpp

```diff
@@ -25,71 +25,95 @@ namespace math {
  * implementation.
  */
 namespace ncr_modulo_p {
+
 /**
- * @brief Class which contains all methods required for calculating nCr mod p
+ * @namespace utils
+ * @brief this namespace contains the definitions of the functions called from
+ * the class NCRModuloP
  */
-class NCRModuloP {
- private:
-    std::vector<uint64_t> fac{};  /// stores precomputed factorial(i) % p value
-    uint64_t p = 0;               /// the p from (nCr % p)
-
- public:
-    /** Constructor which precomputes the values of n! % mod from n=0 to size
-     *  and stores them in vector 'fac'
-     *  @params[in] the numbers 'size', 'mod'
-     */
-    NCRModuloP(const uint64_t& size, const uint64_t& mod) {
-        p = mod;
-        fac = std::vector<uint64_t>(size);
-        fac[0] = 1;
-        for (int i = 1; i <= size; i++) {
-            fac[i] = (fac[i - 1] * i) % p;
-        }
+namespace utils {
+/**
+ * @brief finds the values x and y such that a*x + b*y = gcd(a,b)
+ *
+ * @param[in] a the first input of the gcd
+ * @param[in] a the second input of the gcd
+ * @param[out] x the Bézout coefficient of a
+ * @param[out] y the Bézout coefficient of b
+ * @return the gcd of a and b
+ */
+int64_t gcdExtended(const int64_t& a, const int64_t& b, int64_t& x,
+                    int64_t& y) {
+    if (a == 0) {
+        x = 0;
+        y = 1;
+        return b;
     }
 
-    /** Finds the value of x, y such that a*x + b*y = gcd(a,b)
-     *
-     * @params[in] the numbers 'a', 'b' and address of 'x' and 'y' from above
-     * equation
-     * @returns the gcd of a and b
-     */
-    uint64_t gcdExtended(const uint64_t& a, const uint64_t& b, int64_t* x,
-                         int64_t* y) {
-        if (a == 0) {
-            *x = 0, *y = 1;
-            return b;
-        }
+    int64_t x1 = 0, y1 = 0;
+    const int64_t gcd = gcdExtended(b % a, a, x1, y1);
 
-        int64_t x1 = 0, y1 = 0;
-        uint64_t gcd = gcdExtended(b % a, a, &x1, &y1);
+    x = y1 - (b / a) * x1;
+    y = x1;
+    return gcd;
+}
 
-        *x = y1 - (b / a) * x1;
-        *y = x1;
-        return gcd;
+/** Find modular inverse of a modulo m i.e. a number x such that (a*x)%m = 1
+ *
+ * @param[in] a the number for which the modular inverse is queried
+ * @param[in] m the modulus
+ * @return the inverce of a modulo m, if it exists, -1 otherwise
+ */
+int64_t modInverse(const int64_t& a, const int64_t& m) {
+    int64_t x = 0, y = 0;
+    const int64_t g = gcdExtended(a, m, x, y);
+    if (g != 1) {  // modular inverse doesn't exist
+        return -1;
+    } else {
+        return ((x + m) % m);
     }
+}
+};  // namespace utils
```

Comment:
AFAIK, namespaces don't use `;` at the end of them.
```suggestion
}  // namespace utils
```

Reply:
Ups. Thanks for spotting that.